File: Neetcode Questions/Math & Geometry/powx-n.cpp

// https://leetcode.com/problems/powx-n/

class Solution {
public:

    double solve1(double x, int n){
        if(n == 0) return 1;
        double ans = solve1(x, n/2);
        if(n&1) return (1/x)*ans*ans;
        return ans*ans;
    }

    double solve2(double x, int n){
        if(n == 0) return 1;
        double ans = solve2(x, n/2);
        if(n&1) return x*ans*ans;
        return ans*ans;
    }

    double myPow(double x, int n) {
        if(n < 0) return solve1(x, n);
        return solve2(x, n);
    }
};

File: Neetcode-150-js/Math & Geometry/powx-n.js

// https://leetcode.com/problems/powx-n/

class Solution {
  
  solve1(x, n) {
    if (n === 0) return 1;
    let ans = this.solve1(x, Math.floor(n / 2));
    if (n % 2 === 0) return ans * ans;
    return x * ans * ans;
  }

  solve2(x, n) {
    if (n === 0) return 1;
    let ans = this.solve2(x, Math.floor(n / 2));
    if (n % 2 === 0) return ans * ans;
    return (1 / x) * ans * ans;
  }

  myPow(x, n) {
    if (n === 0) return 1;
    if (n < 0) return this.solve2(x, n);
    return this.solve1(x, n);
  }
}